### PR TITLE
refactor(wasm-builder): wasm builder refactoring and some logic improvements 

### PR DESCRIPTION
**Pull request by grishasobol**
See my comments below in conversations

### CODE REVIEW

File: scripts/src/build.sh [hunk 1]
```diff
@@ -51,9 +51,8 @@ wasm_proc_build() {
 
 # $1 = TARGET DIR
 examples_proc() {
-  # exclude `demo-out-of-memory` because it cannot be processed
   WASM_EXAMPLES_DIR="$1"/wasm32-unknown-unknown/release
-  WASM_EXAMPLES_LIST=$(find $WASM_EXAMPLES_DIR -name "*.wasm" -a -not -name "*demo_out_of_memory*" | tr '\n' ' ' | sed 's/ $//')
+  WASM_EXAMPLES_LIST=$(find $WASM_EXAMPLES_DIR -name "*.wasm" | tr '\n' ' ' | sed 's/ $//')
```
Comment (grishasobol) [inline scripts/src/build.sh:55]: The problem with demo_out_of_memory test is fixed, so return it back

File: utils/wasm-builder/src/wasm_project.rs [hunk 2]
```diff
@@ -261,35 +261,93 @@ extern "C" fn metahash() {{
...
-            .expect("Run `WasmProject::generate()` first");
-
-        let from_path = self
-            .target_dir
-            .join(format!("wasm32-unknown-unknown/{}", self.profile))
-            .join(format!("{}.wasm", &file_base_name));
-
-        fs::create_dir_all(&self.target_dir)?;
-        fs::create_dir_all(&self.wasm_target_dir)?;
+    pub fn postprocess_meta(
```
Comment (grishasobol) [inline utils/wasm-builder/src/wasm_project.rs:264]: split postprocess into to parts: for meta and for opt

File: utils/wasm-builder/src/wasm_project.rs [hunk 3]
```diff
@@ -261,35 +261,93 @@ extern "C" fn metahash() {{
...
-        if !self.project_type.is_metawasm()
-            && smart_fs::copy_if_newer(&from_path, &to_path).context("unable to copy WASM file")?
-        {
-            _ = optimize::optimize_wasm(to_path.clone(), to_path.clone(), "s", false);
+    pub fn postprocess_opt(
+        &self,
+        original_wasm_path: &PathBuf,
+        file_base_name: &String,
+    ) -> Result<()> {
+        let opt_wasm_path = self
```
Comment (grishasobol) [inline utils/wasm-builder/src/wasm_project.rs:300]: Note, now original wasm file and it's copy in `target/wasm32-unknown-unknown/release(debug)/` will never be changed by optimisations, for more clear wasm-builder work.

File: utils/wasm-builder/src/wasm_project.rs (hunk 2 above)
Comment (shamilsan) [inline utils/wasm-builder/src/wasm_project.rs:264]: It would be great to add some docs here.

**Review by shamilsan — APPROVED** [review]
👍

File: utils/wasm-builder/src/wasm_project.rs (hunk 2 above)
Comment (grishasobol) [reply utils/wasm-builder/src/wasm_project.rs:264]: Done

File: utils/wasm-builder/src/wasm_project.rs [hunk 4]
```diff
@@ -261,35 +261,93 @@ extern "C" fn metahash() {{
...
-        fs::create_dir_all(&self.target_dir)?;
-        fs::create_dir_all(&self.wasm_target_dir)?;
+    pub fn postprocess_meta(
+        &self,
+        original_wasm_path: &PathBuf,
+        file_base_name: &String,
+    ) -> Result<()> {
+        let meta_wasm_path = self
+            .wasm_target_dir
+            .join([file_base_name, ".meta.wasm"].concat());
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply utils/wasm-builder/src/wasm_project.rs:271]: Done

**Review by ukint-vs — APPROVED** [review]

**Pull request merged**